DOCSP-46269: atlas search & atlas vector search pages (#3255)
jenssegers/laravel-mongodb: docs/fundamentals/as-avs.txt

Reviewer: Wouldn't it be more readable to use a full name instead of this abbreviations I've never seen before?
`atlas-search.txt`

Author: I wanted to be comprehensive as Atlas Search and Atlas **Vector** Search are two separate features. But I can change the file name to `atlas-search-vector-search`

Reviewer: Maybe create 2 pages that link each other. Indeed that's 2 distinct features.

``$search`` operator reference in the Atlas documentation.

Autocomplete Queries
````````````````````

The {+odm-short+} provides the ``autocomplete()`` method as a query
builder method and as an Eloquent model method. You can use the
``autocomplete()`` method to run autocomplete searches on documents in your
collections.

To learn more about this type of Atlas Search query, see the
:atlas:`autocomplete </atlas-search/autocomplete/>` reference in the
Atlas documentation.

.. note::

   You must create an Atlas Search index with an autocomplete configuration
   on your collection before you can perform autocomplete searches. See the
   :ref:`laravel-as-index` section of this guide to learn more about
   creating Search indexes.

The following code performs an Atlas Search autocomplete query on the
``movies`` collection:

.. io-code-block::
   :copyable: true

   .. input:: /includes/fundamentals/as-avs/AtlasSearchTest.php
      :language: php
      :dedent:
      :start-after: start-auto-query
      :end-before: end-auto-query

   .. output::
      :language: json
      :visible: false

      [
         "Dreaming of Jakarta",
         "Jakob the Liar",

>          "title": "Emily Calling Jake"

Reviewer: Oups, I forgot this prop name.
```suggestion
         "Emily Calling Jake"
```

.. _laravel-atlas-search-avs:

============================
Atlas Search & Vector Search
============================

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: code example, semantic, embeddings

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to perform searches on your documents
by using the Atlas Search and Vector Search features. {+odm-long+}
provides an API to use these features directly with your models.

To learn more about these features, see the following pages in the Atlas
documentation:

- :atlas:`Atlas Search Overview </atlas-search/atlas-search-overview/>`
- :atlas:`Atlas Vector Search Overview </atlas-vector-search/vector-search-overview/>`

.. note:: Deployment Compatibility

   You can use the Atlas Search and Vector Search features only when

>    connecting to MongoDB Atlas clusters. These features are not available for

Reviewer 2: ```suggestion
   You can only use the Atlas Search and Vector Search features when you
   connect to MongoDB Atlas clusters. These features are not available for
```
S: Moved 'only' here for flow and changed tense for simplicity.

Author: Guidance around "only" https://www.mongodb.com/docs/meta/style-guide/terminology/alphabetical-terms/#only:~:text=Don%27t%20hyphenate.-,only,-Place%20only%20preceding

I know it sounds awkward sometimes, but I believe it should stay there

   self-managed deployments.

This page describes how to create the appropriate indexes for each type
of search and provides examples of how to use the {+odm-short+} to
perform searches.

Atlas Search
------------

In this section, you can learn how to create Atlas Search indexes
and perform searches in the {+odm-short+}. The Atlas Search API
internally uses the ``$search`` aggregation operator to perform queries.
To learn more about this operator, see the :atlas:`$search
</atlas-search/aggregation-stages/search/>` reference in the Atlas documentation.

.. _laravel-as-index:

Create an Atlas Search Index
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> .. TODO in DOCSP-46230

Reviewer 2: S: I see you have the ticket for this as well. Maybe comment out the title for the section in the meantime, but up to you depending on your workflow

Author: Fine to leave as it's keeping the references on this page from returning errors atm

Perform Queries
~~~~~~~~~~~~~~~

In this section, you can learn how to use the Atlas Search API in the
{+odm-short+}.

General Queries
```````````````

The {+odm-short+} provides the ``search()`` method as a query
builder method and as an Eloquent model method. You can use the
``search()`` method to run Atlas Search queries on documents in your
collections.

You must pass an ``operator`` parameter to the ``search()`` method that

> is an instance of ``SearchOperatorInterface`` or an array. You can

Reviewer 2: ```suggestion
is an instance of ``SearchOperatorInterface`` or an array of ... . You can
```
Q: An array of what? Would be helpful to include.

create an instance of ``SearchOperatorInterface`` by calling the
``Search::text()`` method and passing the field you are
querying and your search term or phrase.

You must include the following import statement in your application to
create a ``SearchOperatorInterface`` instance:

.. code-block:: php

   use MongoDB\Builder\Search;

> The following code uses the index created in the preceding

Reviewer 2: ```suggestion
The following code uses the ``title`` index created in the preceding
```
S: PLace the name of the index here

Author: Since the index name isn't included in the code example anyways, I'll just rephrase

> The following code uses the index created in the preceding
> :ref:`laravel-as-index` section to perform an Atlas Search query on the

Reviewer 2: ```suggestion
The following code uses the index created in the preceding
:ref:`laravel-as-index` section and a ``SearchOperatorInterface`` instance to perform an Atlas Search query on the
```
S: Mention ``SearchOperatorInterface`` instance, just to tie it back to info in previous paragraph.

Author: I think the info is close enough that readers can infer that the code example includes that class instance

``movies`` collection:

.. io-code-block::
   :copyable: true

   .. input:: /includes/fundamentals/as-avs/AtlasSearchTest.php
      :language: php
      :dedent:
      :start-after: start-search-query
      :end-before: end-search-query

   .. output::
      :language: json
      :visible: false

      [
         { "title": "Dreaming of Jakarta",
           "year": 1990
         },
         { "title": "See You in My Dreams",
           "year": 1996
         }
      ]

You can use the ``search()`` method to perform many types of Atlas
Search queries. Depending on your desired query, you can pass the
following optional parameters to ``search()``:

- ``index``: ``string``
- ``highlight``: ``array``
- ``concurrent``: ``bool``
- ``count``: ``string``
- ``searchAfter``: ``string``
- ``searchBefore`` ``string``
- ``scoreDetails``: ``bool``
- ``sort``: ``array``
- ``returnStoredSource``: ``bool``

> - ``tracking``: ``array``

Reviewer 2: S: Would be good to have a table with a description of what these parameters do. Otherwise, I don't see the point of listing them when readers won't intuitively know what they do.

Author: I outsource these option definitions to the atlas documentation in the link after the list. THis way, we can avoid repeating the same info in two places

Reviewer 2: Hmm, perhaps listing the names of the parameters without the data type would suffice then, since the atlas documentation also includes the data type. The data type seems like unnecessary info if I don't know what the parameter is for. I'll leave up to you though!

builder method and as an Eloquent model method. You can use the
``autocomplete()`` method to run autocomplete searches on documents in your
collections. This method returns only the values of the field you
specify as the query path.

To learn more about this type of Atlas Search query, see the
:atlas:`autocomplete </atlas-search/autocomplete/>` reference in the
Atlas documentation.

.. note::

   You must create an Atlas Search index with an autocomplete configuration
   on your collection before you can perform autocomplete searches. See the
   :ref:`laravel-as-index` section of this guide to learn more about
   creating Search indexes.

The following code performs an Atlas Search autocomplete query for the
string ``"jak"`` on the ``title`` field:

.. io-code-block::
   :copyable: true

   .. input:: /includes/fundamentals/as-avs/AtlasSearchTest.php
      :language: php
      :dedent:
      :start-after: start-auto-query
      :end-before: end-auto-query

   .. output::
      :language: json
      :visible: false

      [
         "Dreaming of Jakarta",
         "Jakob the Liar",
         "Emily Calling Jake"
      ]

> You can also pass the following parameters to the ``autocomplete()``

Reviewer 2: ```suggestion
You can also pass the following optional parameters to the ``autocomplete()``
```
S: For consistency with last section's mention of 'optional' parameters

method to customize the query:

- ``fuzzy``: ``bool`` or ``array`` (default: ``false``)

> - ``tokenOrder``: ``string`` (default: ``'any'``)

Reviewer 2: S: Same here! I would create a table with descriptions of both parameters.

Author: See earlier reply

To learn more about these parameters, see the :atlas:`Options
</atlas-search/autocomplete/#options>` section of the
``autocomplete`` operator reference in the Atlas documentation.

Atlas Vector Search
-------------------

In this section, you can learn how to create Atlas Vector Search indexes
and perform searches in the {+odm-short+}. The Atlas Vector Search API
internally uses the ``$vectorSearch`` aggregation operator to perform queries.
To learn more about this operator, see the :atlas:`$vectorSearch
</atlas-vector-search/vector-search-stage/#mongodb-pipeline-pipe.-vectorSearch>`
reference in the Atlas documentation.

.. _laravel-avs-index:

Create an Atlas Vector Search Index
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. TODO in DOCSP-46230

Perform Queries
~~~~~~~~~~~~~~~

In this section, you can learn how to use the Atlas Vector Search API in
the {+odm-short+}. The {+odm-short+} provides the ``vectorSearch()``
method as a query builder method and as an Eloquent model method. You
can use the ``vectorSearch()`` method to run Atlas Vector Search queries
on documents in your collections.

You must pass the following parameters to the ``vectorSearch()`` method:

> - ``index``: Name of the vector search index (type: ``string``)
> - ``path``: Field that stores vector embeddings (type: ``array`` or ``string``)
> - ``queryVector``: Vector representation of your query (type: ``array``)
> - ``limit``: Number of results to return (type: ``int``)

Reviewer 2: S: The parentheses look a bit weird on the page to me. Consider putting this information in a table or moving the type information by the parameter name.
```suggestion
- ``index`` (type: ``string``) : Name of the vector search index
- ``path`` (type: ``array`` or ``string``): Field that stores vector embeddings 
- ``queryVector`` (type: ``array``): Vector representation of your query 
- ``limit`` (type: ``int``): Number of results to return
```

> The following code uses the index created in the preceding

Reviewer 2: ```suggestion
The following code uses the index created in the preceding
```
S: include the index name

:ref:`laravel-avs-index` section to perform an Atlas Vector Search query on the
``movies`` collection:

.. io-code-block::
   :copyable: true

   .. input::
      :language: php

      $movies = Book::vectorSearch(
          index: 'vector',
          path: 'vector_embeddings',
          // Vector representation of the query `coming of age`
          queryVector: [-0.0016261312, -0.028070757, ...],
          limit: 3,
      );

   .. output::
      :language: json
      :visible: false

      [
         { "title": "Sunrising",
           "plot": "A shy teenager discovers confidence and new friendships during a transformative summer camp experience."
         },
         { "title": "Last Semester",
           "plot": "High school friends navigate love, identity, and unexpected challenges before graduating together."
         }
      ]

You can use the ``vector()`` method to perform many types of Atlas
Search queries. Depending on your desired query, you can pass the
following optional parameters to ``search()``:

- ``exact``: ``bool`` (default: ``false``)
- ``filter``: ``QueryInterface`` or ``array`` (default: no filtering)

> - ``numCandidates``: ``int`` or ``null`` (default: ``null``)

Reviewer 2: S: Same as above. I think I would create a table here with descriptions

.. tip::

   To construct a ``QueryInterface`` instance, you must import the
   ``MongoDB\Builder\Query`` class into your application.

To learn more about these parameters, see the :atlas:`Fields
</atlas-vector-search/vector-search-stage/#fields>` section of the

> ``$vectorSearch`` operator reference in the Atlas documentation.

Reviewer 2: Q: Are API documentation links usually provided for the Laravel docs?

Author: no

</atlas-search/aggregation-stages/search/#fields>` section of the
``$search`` operator reference in the Atlas documentation.

Autocomplete Queries
````````````````````

The {+odm-short+} provides the ``autocomplete()`` method as a query
builder method and as an Eloquent model method. You can use the
``autocomplete()`` method to run autocomplete searches on documents in your
collections.

To learn more about this type of Atlas Search query, see the
:atlas:`autocomplete </atlas-search/autocomplete/>` reference in the
Atlas documentation.

.. note::

   You must create an Atlas Search index with an autocomplete configuration
   on your collection before you can perform autocomplete searches. See the
   :ref:`laravel-as-index` section of this guide to learn more about
   creating Search indexes.

The following code performs an Atlas Search autocomplete query on the
``movies`` collection:

.. io-code-block::
   :copyable: true

   .. input:: /includes/fundamentals/as-avs/AtlasSearchTest.php
      :language: php
      :dedent:
      :start-after: start-auto-query
      :end-before: end-auto-query

   .. output::
      :language: json
      :visible: false

>       [
>          { "title": "Dreaming of Jakarta",
>            "year": 1990
>          },
>          { "title": "Jakob the Liar",
>            "year": 1999
>          },
>          { "title": "Emily Calling Jake",
>            "year": 2001
>          }
>       ]

Reviewer: This method returns the list of strings for autocompletion, not Model instance.

```suggestion
      [
         "Dreaming of Jakarta",
         "Jakob the Liar",
         "title": "Emily Calling Jake"
      ]
```

.. _laravel-as-index:

Create an Atlas Search Index
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. TODO in DOCSP-46230

Perform Queries
~~~~~~~~~~~~~~~

In this section, you can learn how to use the Atlas Search API in the
{+odm-short+}.

General Queries
```````````````

The {+odm-short+} provides the ``search()`` method as a query
builder method and as an Eloquent model method. You can use the
``search()`` method to run Atlas Search queries on documents in your
collections.

You must pass an ``operator`` parameter to the ``search()`` method that
is an instance of ``SearchOperatorInterface`` or an array that contains
the operator type, field name, and query value. You can
create an instance of ``SearchOperatorInterface`` by calling the
``Search::text()`` method and passing the field you are
querying and your search term or phrase.

You must include the following import statement in your application to
create a ``SearchOperatorInterface`` instance:

.. code-block:: php

   use MongoDB\Builder\Search;

The following code performs an Atlas Search query that searches for the
term ``'dream'`` in the ``title`` field values of documents in the

> ``movies`` collection:

Reviewer 2: ```suggestion
The following code performs an Atlas Search query in the
``movies`` collection on the ``title`` field for the term ``'dream'``:
```
S: Possible rewording for conciseness

.. io-code-block::
   :copyable: true

   .. input:: /includes/fundamentals/as-avs/AtlasSearchTest.php
      :language: php
      :dedent:
      :start-after: start-search-query
      :end-before: end-search-query

   .. output::
      :language: json
      :visible: false

      [
         { "title": "Dreaming of Jakarta",
           "year": 1990
         },
         { "title": "See You in My Dreams",
           "year": 1996
         }
      ]

You can use the ``search()`` method to perform many types of Atlas
Search queries. Depending on your desired query, you can pass the
following optional parameters to ``search()``:

.. list-table::
   :header-rows: 1

>    * - Parameter

Reviewer 2: ```suggestion
   * - Optional Parameter
```
S: for clarity

     - ``int``
     - Number of results to return

The following code uses the ``vector`` index created in the preceding
:ref:`laravel-avs-index` section to perform an Atlas Vector Search query on the
``movies`` collection:

.. io-code-block::
   :copyable: true

   .. input::
      :language: php

      $movies = Book::vectorSearch(
          index: 'vector',
          path: 'vector_embeddings',
          // Vector representation of the query `coming of age`
          queryVector: [-0.0016261312, -0.028070757, ...],
          limit: 3,
      );

   .. output::
      :language: json
      :visible: false

      [
         { "title": "Sunrising",
           "plot": "A shy teenager discovers confidence and new friendships during a transformative summer camp experience."
         },
         { "title": "Last Semester",
           "plot": "High school friends navigate love, identity, and unexpected challenges before graduating together."
         }
      ]

You can use the ``vector()`` method to perform many types of Atlas
Search queries. Depending on your desired query, you can pass the
following optional parameters to ``search()``:

.. list-table::
   :header-rows: 1

>    * - Parameter

Reviewer 2: ```suggestion
   * - Optional Parameter
```
